### PR TITLE
feat: item pools — constant-product AMM over fungible items

### PR DESCRIPTION
**Pull request by JirAcheron**
Constant-product (x·y=k) pools over any two fungible items. MUSU is item index 1, so item↔currency pools need no special casing. Ops-only creation/seeding; players swap and LP; fees accrue to reserves (LPs earn pro-rata on burn).

**Contracts**: LibPool + LibPoolRegistry (Listing-convention registry/runtime split), PoolSystem (swap/addLiquidity/removeLiquidity), _PoolRegistrySystem (create/donate/setFee/setDisabled/remove). Pool entity holds fee in Rate, LP supply in Value; reserves are pool-owned inventories; LP shares are POOL_SHARE entities with the seed √(a·b) locked to the pool itself. Zero new component types. All rounding floors in the pool's favor — k strictly increases per swap.

**Client**: shapes/Pool (integer-exact pricing mirror), api.player.pool, two-tab modal (Swap / Liquidity) via the Trading menu.

**Verification**: 21 forge tests (creation/permissions/slippage/fee-accrual/locked-liquidity/pause/removal/MUSU pair); on-chain smoke script (deployment/contracts/PoolSmoke.s.sol) run against a fresh local world; client typecheck clean.

Full design/domain/math writeup: game-kamigotchi/docs/item-pools-report.html

## Summary by CodeRabbit

* **New Features**
  * Added an “Item Pools” menu option with a dedicated pool modal.
  * Enables swapping between two pool items and managing liquidity (add/remove LP shares) with live reserve/account refresh and pool selection.
* **Bug Fixes**
  * Improved swap and liquidity safety with slippage-protected minimums, tighter validation, and clearer disabled-pool behavior (paused indicator; swap/add blocked, exits still allowed).
  * Refined AMM/pricing math to align with on-chain rounding.
* **Tests**
  * Expanded pool system test coverage for creation, swaps, liquidity, fee/disable controls, and edge cases.

### CODE REVIEW

File: packages/contracts/src/libraries/LibPoolRegistry.sol [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+  /// @notice remove all data associated with a pool
+  /// @dev reserves and locked shares must be cleared beforehand (see _PoolRegistrySystem)
+  function remove(IUintComp comps, uint256 id) internal {
+    LibEntityType.remove(comps, id);
+    KeysComponent(getAddrByID(comps, KeysCompID)).remove(id);
+    RateComponent(getAddrByID(comps, RateCompID)).remove(id);
+    ValueComponent(getAddrByID(comps, ValueCompID)).remove(id); // total LP supply
+    TimeStartComponent(getAddrByID(comps, TimeStartCompID)).remove(id);
+    LibDisabled.set(comps, id, false);
```
[the remark this answers is not in the document]
Comment (JirAcheron) [reply packages/contracts/src/libraries/LibPoolRegistry.sol]: Checked against the implementation: `LibDisabled.set(comps, id, false)` calls `IsDisabledComponent.remove(id)` — the false branch removes the entry outright rather than writing false, so nothing stale survives removal and recreated pairs start enabled. Added a clarifying comment at the call site. The other three findings are addressed in 8501b5754 (512-bit mulDiv, V2-style assert, BigInt client pricing).

**Comment by JirAcheron** [comment]
@greptileai review the latest changes please — review feedback addressed.

File: packages/contracts/src/systems/_PoolRegistrySystem.sol [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    require(indexA != indexB, "Pool: identical items");
+    require(LibItem.getByIndex(components, indexA) != 0, "Item: does not exist");
+    require(LibItem.getByIndex(components, indexB) != 0, "Item: does not exist");
+    verifyTradable(indexA, indexB);
+    require(feeBps <= MAX_FEE_BPS, "Pool: fee too high");
+    require(amtA >= MINIMUM_SEED && amtB >= MINIMUM_SEED, "Pool: seed too small");
+    require(LibPoolRegistry.get(components, indexA, indexB) == 0, "Pool already exists");
+
+    id = LibPoolRegistry.create(components, indexA, indexB, feeBps);
+    LibInventory.incFor(components, id, indexA, amtA);
```
Comment (kanyaveral) [inline packages/contracts/src/systems/_PoolRegistrySystem.sol]: **blocking: the MUSU/ONYX pool cannot be created.**

`incFor` calls `LibItem.verifyToken(comps, itemIndex, false)`, which reverts for any item carrying a TokenAddress. Onyx Shard is item index 100, type `ERC20`. so this line reverts with `LibItem: item has a token` for the exact pair this feature exists to ship. `donate()` (line 55) has the same problem.

reproduced against this branch:
```
[FAIL: LibItem: item has a token] testCreateMusuOnyxPool()
[PASS] testCreateMusuPlainPool()      <- identical call, plain item instead of ONYX
```

the guard is correct, so this is not a one-liner: minting a token-backed item from thin air would create shards with no ERC20 backing in the portal's TokenHolder. seeding an ONYX pool properly means transferring already-bridged shards from a treasury account (via `transferFor`, which does permit token items) rather than minting them. `remove()` then also needs a treasury-return path, since it currently burns reserves via `decFor` and would destroy backed shards while stranding the underlying ONYX.

at minimum, make the invariant explicit here with a `require` rather than leaving it as an incidental side effect of the faucet implementation: a future refactor of seeding would silently make ONYX pools creatable, and the runtime paths (`transferFor`) do allow token items.

File: packages/contracts/src/systems/_PoolRegistrySystem.sol [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    LibInventory.incFor(components, id, indexB, amtB);
+    LibPool.mintShares(components, id, id, LibPool.calcInitialLiquidity(amtA, amtB));
```
Comment (kanyaveral) [inline packages/contracts/src/systems/_PoolRegistrySystem.sol:61]: **blocking: a griefer can hijack the pool's launch price.**

two things combine here. `ItemTransferSystem.execute` never validates that `targetID` is an account, and `LibPoolRegistry.genID` is a pure hash, so anyone can compute a pool's address and fund it *before it exists*. and this line mints locked shares as `sqrt(amtA * amtB)` from the **arguments**, not from the actual post-seed reserves.

reproduced: attacker sends 1,000,000 of itemA to the not-yet-created poolID, then admin seeds an intended 1:1 pool at 100k/100k:
```
reserveA (admin intended 100000): 1,100,000
reserveB (admin intended 100000):   100,000
locked shares = sqrt(arg*arg):      100,000
```
the pool launches at 11:1 instead of 1:1, handing free arbitrage to whoever is watching.

cheap fix: `require` both reserves are empty before seeding (`getBalanceOf(...) == 0` for lo and hi), or compute the seed shares from real balances after the incs. the first is better: it also surfaces the griefing attempt instead of silently absorbing it.

File: packages/contracts/src/libraries/LibPool.sol [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+    require(amountIn > 0, "Pool: zero input");
+
+    uint256 reserveIn = LibInventory.getBalanceOf(comps, poolID, indexIn);
+    uint256 reserveOut = LibInventory.getBalanceOf(comps, poolID, indexOut);
+    amountOut = calcAmountOut(amountIn, reserveIn, reserveOut, LibRegistry.getFeeBps(comps, poolID));
+    require(amountOut > 0, "Pool: insufficient output");
+    require(amountOut >= minAmountOut, "Pool: slippage exceeded");
+
+    LibInventory.transferFor(comps, accID, poolID, indexIn, amountIn); // implicit balance check
+    LibInventory.transferFor(comps, poolID, accID, indexOut, amountOut);
```
Comment (kanyaveral) [inline packages/contracts/src/libraries/LibPool.sol]: **blocking: swap payouts inflate the MUSU leaderboard.**

`transferFor` -> `_incFor` calls `logItemTotal`, which credits the receiving account's `ITEM_TOTAL` for `indexOut`. `ITEM_TOTAL[MUSU]` is exactly what kamiden's "Account Musu" leaderboard reads (`pkg/indexer/kamiden/leaderboard.go`, `getScoreByAccount("1", "ITEM_TOTAL")`) and what the client shows as coin earned (`network/shapes/Account/stats.ts`). so a pool payout of MUSU reads as *earning* MUSU.

reproduced, one account round-tripping itemA -> MUSU -> itemA ten times:
```
ITEM_TOTAL(MUSU) after 10 round-trips: 98,684
alice actual MUSU balance:                   0
```

no counterparty needed, ~0.6% in fees per cycle. unlike listing buys (a full-price sink) or P2P transfers (counterparty + MUSU fee), this is solo and unbounded. and it fires on honest use too, since swapping items for MUSU counts as earning it, so pool volume structurally corrupts the stat.

suggest the swap legs bypass the acquisition logging (a `transferFor` variant that skips `logItemTotal`, or move pool payouts onto `_decFor`/`_incFor` directly and log pool-specific data instead). `POOL_VOLUME` in `logSwap` already covers the analytics need.

File: packages/contracts/src/libraries/LibPool.sol [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+    uint256 poolID,
+    uint256 accID,
+    RemoveLiqParams memory p
+  ) public returns (uint256 amtA, uint256 amtB) {
+    require(p.shares > 0, "Pool: zero shares");
+
+    uint256 supply = getTotalSupply(comps, poolID);
+    amtA = FixedPointMathLib.mulDivDown(p.shares, LibInventory.getBalanceOf(comps, poolID, p.indexA), supply);
+    amtB = FixedPointMathLib.mulDivDown(p.shares, LibInventory.getBalanceOf(comps, poolID, p.indexB), supply);
+    require(amtA > 0 && amtB > 0, "Pool: insufficient liquidity burned");
```
Comment (kanyaveral) [inline packages/contracts/src/libraries/LibPool.sol]: **blocking: small LP positions become permanently unexitable.**

requiring `amtA > 0 && amtB > 0` is safe at 1e18 wei scale (Uniswap's same guard is unreachable there), but these are game-scale integers. a small position in a ratio-skewed pool floors one side to zero and can never be burned:
```
pool seeded 1,000,000 : 1,000
alice adds 1,000 A + 10 B  ->  31 shares
her B slice = 31 * rB / supply = 0
-> removeLiquidity(31) reverts "Pool: insufficient liquidity burned"
```
her whole deposit is frozen until the ratio moves, and every sub-amount fails too. an attacker can also induce this by skewing reserves with a swap.

`MINIMUM_SEED = 1000` ("keeps integer rounding tolerable") shows the class was considered, but it only bounds the seed, not LP positions. options: require only that `amtA + amtB > 0` and let a zero side be a zero side, or enforce a minimum share position on add so a stake that cannot be exited cannot be created. the client makes it worse by rendering the remove button with no warning.

File: packages/client/src/app/components/modals/pool/Pool.tsx [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+  calcAmountOut,
+  calcRemoveAmounts,
+  calcSharesMinted,
+  getAllPools,
+  getPoolShares,
+  Pool,
+  quote,
+} from 'network/shapes/Pool';
+
+const SLIPPAGE_BPS = 100; // 1% tolerance on swaps and liquidity adds
```
Comment (kanyaveral) [inline packages/client/src/app/components/modals/pool/Pool.tsx:34]: **slippage protection is a fiction on small trades.**

`applySlippage` floors, so at game-scale quotes the 1% bound collapses:
- quote of 2 -> `minOut = floor(2 * 9900 / 10000) = 1` (50% tolerance)
- quote of 1 -> `minOut = 0` (unbounded; only `amountOut > 0` on-chain)

anything below ~100 units gets materially less protection than the "1%" the UI displays. Yominet is FCFS with a visible mempool, so this is a live sandwich surface, and there is no deadline parameter either.

suggest surfacing the actual computed `minOut` in the UI rather than a nominal percentage, and considering a floor of at least 1 unit of protection (or blocking swaps whose quote is too small to protect).

File: packages/client/src/app/components/modals/pool/Pool.tsx [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+    }, []);
+
+    // refresh pools + account on each tick while open
+    useEffect(() => {
+      if (!poolModalOpen) return;
+      setPools(getAllPools(world, components));
+      setAccount(_getAccount(world, components, accountEntity, { live: 2, inventory: 2 }));
+    }, [lastTick, poolModalOpen, accountEntity]);
+
+    const pool = useMemo(
```
Comment (kanyaveral) [inline packages/client/src/app/components/modals/pool/Pool.tsx]: **stale pool selection can execute against the wrong pair.**

`pools.find(...) ?? pools[0]` silently falls back when the stored `poolID` is no longer in the list, but `swapInput` / `addAmountA` / `removeShares` are only reset in the select's `onChange` (line 269). so if the pool list changes while the modal is open (pool removed or disabled mid-session), `pool` flips to a different pair while the typed amount persists, and a click submits that amount against an item the user never chose.

reset the amounts in an effect keyed on `pool?.id` rather than only in `onChange`.

File: packages/client/src/app/components/modals/pool/Pool.tsx [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+          </Row>
+          <Row>
+            <Input
+              type='number'
+              min='0'
+              value={swapInput || ''}
+              placeholder='0'
+              onChange={(e) => setSwapInput(parseAmount(e.target.value))}
+            />
+            <FlipButton onClick={() => setInverted(!inverted)}>⇅</FlipButton>
```
Comment (kanyaveral) [inline packages/client/src/app/components/modals/pool/Pool.tsx]: **flip does not reset the input, so the amount silently re-denominates.**

user types `1000` meaning 1000 MUSU, taps the arrow to check the reverse rate, hits swap: the same `1000` now means 1000 of the counter-item, potentially a very different value per unit. the balance check can still pass. suggest `setSwapInput(0)` on flip (or converting the amount at the current rate).

File: packages/contracts/src/systems/_PoolRegistrySystem.sol [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    LibPoolRegistry.setDisabled(components, id, disabled);
+  }
+
+  /// @notice delete a pool once all players have exited, burning remaining reserves
+  function remove(uint32 indexA, uint32 indexB) public onlyAdmin(components) {
+    uint256 id = LibPoolRegistry.get(components, indexA, indexB);
+    require(id != 0, "Pool does not exist");
+
+    uint256 locked = LibPool.getShares(components, id, id);
+    require(LibPool.getTotalSupply(components, id) == locked, "Pool: shares outstanding");
```
Comment (kanyaveral) [inline packages/contracts/src/systems/_PoolRegistrySystem.sol]: **one dust LP share blocks pool removal forever.**

`totalSupply == locked` means any player holding a single share can prevent deletion indefinitely. `setDisabled` still lets them hold, since exits are permitted but never compelled. the pair slot is a deterministic hash, so the pool can never be re-parameterized via delete/recreate.

not urgent (no value is extractable), but worth an admin force-exit path that burns remaining shares to their holders' inventories.

File: packages/contracts/src/systems/PoolSystem.sol [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+    LibPool.logLiquidity(world, components, accID, poolID, true, amtA, amtB, liquidity);
+
+    // standard logging and tracking
+    LibAccount.updateLastTs(components, accID);
+
+    return (amtA, amtB, liquidity);
+  }
+
+  /// @notice burn LP shares for a pro-rata slice of both reserves
+  function removeLiquidity(
```
Comment (kanyaveral) [inline packages/contracts/src/systems/PoolSystem.sol:89]: **design: bundling `removeLiquidity` with `swap` weakens the exit guarantee.**

the file comment says "Liquidity removal deliberately works on disabled pools so LPs can always exit", and at the entity level that holds. but the guarantee is only as strong as the system's availability: if a bug is found in `swap` or `addLiquidity`, the standard solecs response is to deregister/replace `PoolSystem`, which takes `removeLiquidity` down with it and freezes every LP position exactly when it matters most.

the codebase's own convention points the other way: `TradeCancelSystem` is separate from `TradeExecuteSystem` for precisely this reason, and listings split buy/sell. a separate `PoolExitSystem` would keep the exit path independently deployable.

File: packages/client/src/network/shapes/Pool/types.ts [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+  holderID: EntityID
+): number => {
+  const entity = genShareEntity(world, poolID, holderID);
+  return entity ? getValue(comps, entity) : 0;
+};
+
+/////////////////
+// ENTITY HASHES (mirror LibPool/LibPoolRegistry genIDs)
+
+export const genReserveEntity = (world: World, poolID: EntityID, itemIndex: number) => {
```
Comment (kanyaveral) [inline packages/client/src/network/shapes/Pool/types.ts]: **cleanup: this duplicates existing Inventory helpers.**

`queryInventoryInstance` (`network/shapes/Inventory/queries.ts`) already hashes `['inventory.instance', holderID, itemIndex]`, and `getItemBalance` (`network/shapes/Item/functions.ts`) already resolves that entity to a Value balance. `getReserve` can just call `getItemBalance(world, comps, poolID, itemIndex)`.

worth doing because it is a duplicated cross-layer contract: if the inventory ID scheme ever changes, the Inventory shape gets updated and this copy silently keeps hashing the old preimage, making every reserve read 0.

also in this file: `get()` uses the uncached `getItemByIndex`, so the modal's 1s tick rebuilds full Item shapes (requirements/effects traversals) for every pool every second. `app/cache/item`'s `getByIndex` is a permanent cache and would make it a map lookup.

File: packages/client/src/app/components/modals/pool/Pool.tsx [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+    const [inverted, setInverted] = useState(false);
+    const [swapInput, setSwapInput] = useState(0);
+
+    // liquidity state
+    const [addAmountA, setAddAmountA] = useState(0);
+    const [removeShares, setRemoveShares] = useState(0);
+
+    // ticking
+    useEffect(() => {
+      const timerID = setInterval(() => setLastTick(Date.now()), 1000);
```
Comment (kanyaveral) [inline packages/client/src/app/components/modals/pool/Pool.tsx:119]: **cleanup: this interval ticks forever, including while the modal is closed.**

`MainWindow` renders `allComponents` unconditionally, so `PoolModal` is permanently mounted and this interval keeps firing for the rest of the session. the refresh effect below is gated on `poolModalOpen`, but the render body (`pools.find`, `getPoolShares` with its keccak, the BigInt pricing math) still re-runs every second with the modal invisible.

`KamiSend.tsx` and marketplace `Bids.tsx` show the cheaper pattern: create the interval inside an effect gated on the open state, so a closed modal costs nothing.

also, this modal hand-rolls `ActionButton`, `TabButton` and `Title` as local styled-components while `app/components/library` already exports `ActionButton` (with `playClick()` SFX and tooltip support, used by 10+ modals) and `ModalHeader` (used by essentially every other modal). the local clones shadow the library names while dropping the behavior, so pool ends up the one modal with no click sound and a non-standard header.

File: packages/contracts/test/systems/Pool.t.sol [hunk 13]
```diff
@@ -0,0 +1,454 @@
...
+    // same pair can be recreated
+    uint256 newID = _createPool();
+    assertEq(newID, poolID);
+    assertEq(_getItemBal(poolID, ITEM_A), SEED_A);
+  }
+
+  /////////////////
+  // MUSU PAIR
+
+  function testMusuPair() public {
```
Comment (kanyaveral) [inline packages/contracts/test/systems/Pool.t.sol:474]: **test gap: no ERC20-backed item case.**

`testMusuPair` covers the currency pairing, but every test in this suite uses plain items or MUSU, which is why the ONYX blocker (see `_PoolRegistrySystem.sol:41`) went unnoticed. worth adding a case that attaches a token via `_addItemERC20` (the helper already exists in `SetupTemplate.t.sol`) and asserts whatever the intended behavior is: either an explicit revert with a clear message, or a working treasury-seeded pool once that path exists.

a couple of other cases worth pinning down while you are here: a small-LP-position exit in a skewed pool (currently reverts, see `LibPool.sol:128`), and `create()` on a pool ID that already holds a pre-seeded balance (see `_PoolRegistrySystem.sol:43`).

**Review by kanyaveral — CHANGES_REQUESTED** [review]
## audit summary

reviewed at head `8501b57`. ran the suite locally: **21/21 pool tests pass**, full contracts suite is 365 passing with 16 failures that I confirmed **pre-exist on `main`** (all `testMintTicket*` fuzz cases, unrelated to pools). CI is green.

the AMM core is well built and worth keeping. it's a faithful V2 port that fits the ECS grammar with zero new component types, and the classic attacks are properly defended: first-depositor share inflation is precluded by admin-only seeding with locked seed shares, `donate` is admin-gated, `getByOperator` reverts for non-accounts, LP shares are non-transferable, and k never decreases even at `feeBps=0`. the bugs below are at the edges, not in the math.

**four blocking issues, each reproduced with a test written against this branch.**

### 1. the MUSU/ONYX pool cannot be created (blocking)

`create()` seeds via `LibInventory.incFor`, which calls `LibItem.verifyToken(index, false)` and reverts for any item carrying a TokenAddress. Onyx Shard is item index 100, type `ERC20`. probe:

```
[FAIL: LibItem: item has a token] testCreateMusuOnyxPool()
[PASS] testCreateMusuPlainPool()      <- identical call, plain item instead of ONYX
```

`donate()` uses the same `incFor` path, so it is blocked too. the pool tests only ever use plain items and MUSU, so nothing caught it.

the guard is *correct*, which is what makes this substantive rather than a one-liner: minting a token-backed item from thin air would create shards with no ERC20 backing behind them in the portal's TokenHolder. seeding an ONYX pool properly means transferring already-bridged shards from a treasury account (via `transferFor`, which does permit token items) instead of minting them, and then `remove()` (which burns reserves via `decFor`) needs a treasury-return path or it destroys backed shards and strands the underlying ONYX. worth designing before more code lands on this stack.

### 2. swap payouts inflate the MUSU leaderboard (blocking)

`transferFor`'s `_incFor` calls `logItemTotal`, crediting the receiver's `ITEM_TOTAL` on every swap. `ITEM_TOTAL[MUSU]` is exactly kamiden's "Account Musu" leaderboard and the client's coin-earned stat, so a pool payout of MUSU reads as *earning* MUSU. ten round-trips:

```
ITEM_TOTAL(MUSU) after 10 round-trips: 98,684
alice actual MUSU balance:                   0
```

solo, no counterparty, costing only ~0.6% in fees per cycle. any MUSU pool turns the earned-MUSU leaderboard into a wash-trading contest, and this fires on honest use too since swapping items for MUSU counts as earning it.

### 3. anyone can hijack a pool's launch price (blocking)

`ItemTransferSystem.execute` never validates that `targetID` is an account, and `genID` is a pure hash, so a griefer can compute the pool's address and fund it *before it exists*. `create()` then mints locked shares as `sqrt(argA*argB)` from its arguments rather than from actual post-seed reserves:

```
reserveA (admin intended 100000): 1,100,000
reserveB (admin intended 100000):   100,000
locked shares = sqrt(arg*arg):      100,000
```

the admin seeds an intended 1:1 pool and it launches at 11:1, handing free arbitrage to whoever is watching.

### 4. small LP positions become permanently unexitable (blocking)

`removeLiquidity` requires `amtA > 0 && amtB > 0`. at game-scale integers (not 1e18 wei, where this guard is unreachable), a small position in a ratio-skewed pool floors one side to zero:

```
alice shares: 31
her B slice = shares*rB/supply: 0
-> removeLiquidity reverts "Pool: insufficient liquidity burned"
```

her deposit is frozen until the ratio moves. the `MINIMUM_SEED = 1000` comment shows awareness of this class but it doesn't cover LP positions, and the client renders the remove button with no warning.

### should-fix

- **client slippage is a fiction on small trades.** `SLIPPAGE_BPS=100` floors: a quote of 2 gives `minOut=1` (50% tolerance), a quote of 1 gives `minOut=0` (unbounded). the UI says "1%". on Yominet's FCFS public mempool this is a live sandwich surface, and there is no deadline parameter.
- **stale pool selection** can send a queued swap against a different item pair.
- **flip button does not reset the input**, silently re-denominating the amount.
- **one dust LP share blocks pool removal forever.**

### design questions

- **untaxed barter channel.** `LibTrade` deliberately forbids barter ("Please use MUSU") and charges MUSU fees plus tax. any item/item pool reintroduces barter, and any MUSU pool is an untaxed venue that undercuts taxed P2P trades and guts a MUSU sink. worth a deliberate call rather than an emergent one.
- **bundling swap/add/remove in one system** breaks the stated "LPs can always exit" guarantee: hot-patching a swap bug means deregistering `PoolSystem`, which takes `removeLiquidity` with it. the codebase's own convention (`TradeCancelSystem` separate from `TradeExecuteSystem`) exists for exactly this reason.

### cleanup (non-blocking)

`getReserve` duplicates the existing Inventory shape helpers; the modal hand-rolls `ActionButton`/`ModalHeader` instead of the shared library (losing click SFX and the standard header); `verifyTradable` is copy-pasted across both systems; the `get` + `require` pair repeats 7x; `LibPool.min` duplicates solady's; the 1s interval ticks forever even while the modal is closed; `getItemByIndex` rebuilds full Item shapes every second uncached; 5 files fail the repo's prettier config.

inline comments below for each.

**Comment by JirAcheron** [comment]
Thanks @kanyaveral — reproduced all of these, they're real. Fixes in `ec2d8df3f` (contracts +4 tests) and `a5c270594` (client):

**Fixed (blocking):**
- **Launch-price grief** — `create` now `require`s both reserves empty before seeding (pool id is a pure hash, so pre-funding was possible). Test `testCreateRejectsPreFundedReserves`.
- **Token-backed items** — explicit `verifyNoToken` guard in `create`/`donate` (no longer incidental to `incFor`). Test `testCreateRejectsTokenItem`.
- **Unexitable dust LP** — remove guard is now `amtA > 0 || amtB > 0`; a floored-to-zero side is forfeited rather than freezing the position. Test `testRemoveDustSideStillExits`.
- **MUSU leaderboard inflation** — added `LibInventory.transferForNoLog` (no `ITEM_TOTAL`/`ITEM_COUNT` write) and routed all pool↔account legs through it, so swap/LP payouts no longer read as earned MUSU. Test `testSwapDoesNotInflateItemTotal`.

**Fixed (client):** tick interval now gated on modal-open; amounts reset on pool change **and** flip; the UI shows the real enforced `minReceived` and blocks trades too small for the 1% floor to protect (your `quote → 0` sandwich surface).

**Deferred — need @Acheron's mechanism-design call, not landing these unilaterally on a live economy:**
- **Full ONYX pool support.** The guard makes exclusion explicit and safe today. Real support needs the treasury seed/return model you described (`transferFor` from a bridged-shard treasury, and a treasury-return path in `remove` instead of `decFor` burning backed shards). That's a product decision + its own PR.
- **Dust LP share blocks pool `remove` forever** (`_PoolRegistrySystem:79`) and **bundling `removeLiquidity` with `swap`** (`PoolSystem:89`) — both are design tradeoffs (griefability of pool teardown vs. per-action systems). Flagging for a design call rather than picking one.
- **`types.ts:66` duplicates Inventory helpers** and **`Pool.t.sol` ERC20 test gap** — the ERC20 case is now covered by `testCreateRejectsTokenItem`; the helper de-dup is a low-risk cleanup I can fold in if you'd like.

Not merging #2425 until you've weighed in on the deferred items.

**Comment by JirAcheron** [comment]
@greptileai re-review please — all blocking findings from the review addressed (see summary above). Full suite green: 25 pool tests + auth escalation test.

File: packages/client/src/network/shapes/Pool/types.ts (hunk 11 above)
Comment (JirAcheron) [reply packages/client/src/network/shapes/Pool/types.ts]: Fixed in db064d052: getReserve now delegates to getItemBalance (single inventory-ID source, genReserveEntity removed), and get() uses the cached getItemByIndex from app/cache/item so the 1s tick is a map lookup rather than a full Item-shape rebuild.

File: packages/contracts/src/systems/PoolSystem.sol (hunk 10 above)
Comment (JirAcheron) [reply packages/contracts/src/systems/PoolSystem.sol:89]: Decision: keeping the three actions bundled in one PoolSystem for now (product call). The exit guarantee holds at the function level (removeLiquidity skips pause/tradable checks); the residual risk is availability — deprecating/replacing PoolSystem during an incident or a botched upgrade takes the exit path down with swap/add. Accepted given standard upgrade hygiene (never deprecate without a live replacement). If we later need to circuit-break swaps during incidents, splitting out a standalone PoolLiquidityRemoveSystem is the cheap follow-up. Resolving as won't-fix-by-design.

File: packages/contracts/src/systems/_PoolRegistrySystem.sol (hunk 2 above)
Comment (JirAcheron) [reply packages/contracts/src/systems/_PoolRegistrySystem.sol]: Resolved by reworking the seeding model (afbae1b41): pools no longer mint reserves from thin air — create/donate transfer the seed from the calling admin's own inventory (transferForNoLog), so token-backed items (ONYX) are supported with real, backed shards. The token guard is removed, and remove() returns the residual seed to the admin instead of burning (so backed shards are never destroyed). Tests: testCreateAllowsTokenItem (ONYX pool creatable), testCreateRevertsIfSeederShort, testRemovePool asserts the seed returns to the admin. The admin/treasury must be a game account holding the items (per your call to seed from the caller's account). 27 pool tests green.

**Comment by JirAcheron** [comment]
@greptileai review

**Comment by kanyaveral** [comment]
@greptileai review last commit

File: packages/contracts/src/systems/_PoolRegistrySystem.sol [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+    require(LibPoolRegistry.get(components, indexA, indexB) == 0, "Pool already exists");
+
+    // the pool id is a pure hash of the item pair, so anyone can precompute it
+    // and pre-fund the entity before creation (ItemTransferSystem does not
+    // require an account target). sweep any pre-funded balance to the caller:
+    // reverting instead would let a 1-unit deposit permanently brick the pair
+    // (no donate/remove path exists pre-creation), and seeding on top would
+    // launch the pool at a skewed price. sweeping keeps the launch price at
+    // exactly the seeded ratio, with locked shares = sqrt of the true reserves.
+    id = LibPoolRegistry.genID(indexA, indexB);
```
[the remark this answers is not in the document]
Comment (kanyaveral) [reply packages/contracts/src/systems/_PoolRegistrySystem.sol:48]: not valid: `genID` sorts internally (`LibPoolRegistry.sol:104-107` calls `sortIndices` before hashing), so the sweep, the reserve checks and the created pool all resolve to the same id regardless of argument order. `testPoolCreate` already pins order-insensitive lookup (`get(ITEM_B, ITEM_A) == poolID`). the explicit `sortIndices` in `LibPoolRegistry.create` exists to build `Keys = [lo, hi]`, not to compensate for `genID`. the proposed change is a no-op.

**Comment by kanyaveral** [comment]
@greptileai check this out again

**Comment by kanyaveral** [comment]
@greptileai review again - focus on UI changes

**Review by kanyaveral — DISMISSED** [review]
think we can send it

**Review by kanyaveral — DISMISSED** [review]
approved and standing by

**Review by kanyaveral — DISMISSED** [review]

**Review by kanyaveral — APPROVED** [review]
